### DISCUSSION TITLE
Is there a way to configure `NO_LIMIT` in `org.janusgraph.graphdb.query.BaseQuery`

### DISCUSSION THREAD

**Discussion by sauvermazeotap** [JanusGraph/janusgraph#2581] [Q&A] [answered]
- NO_LIMIT is defined as Integer.MAX_VALUE (2147483647) in org.janusgraph.graphdb.query.Query
- BaseQuery (parent of all query classes) is using NO_LIMIT for slice queries
- On K,V backends, this can get into full scan query
- Is there a way to configure `BaseQuery` limit at runtime? Also why is it set to Integer.MAX_VALUE?

**Comment by li-boxuan** [comment, marked as answer]
Can you give a scenario where you want to configure the limit? Does [gremlin limit step](https://tinkerpop.apache.org/docs/current/reference/#limit-step) not work for you?

**Reply by sauvermazeotap** [reply]
Thanks @li-boxuan for the quick revert.

I would prefer to use the gremlin `limit` step, please shed light on below observation

g.V().has('idValue','011c40f6-18d8-4111-77be-7ff5238d1eed').has('idType','id_mid_4').limit(1).bothE().otherV().profile()

**Traversal Metrics**
Step                                                               Count  Traversers       Time (ms)    % Dur
JanusGraphStep([],[idValue.eq(011c40f6-18d8-411...                     1           1           4.337    39.66
    \_condition=(idValue = 011c40f6-18d8-4111-77be-7ff5238d1eed AND idType = id_mid_4)
    \_isFitted=true
    \_query=multiKSQ[1]@2147483647
    **\_limit=1**
    \_index=idIndex
    \_orders=[]
    \_isOrdered=true
  optimization                                                                                 0.023
  optimization                                                                                 0.216
  backend-query                                                                                0.000
    \_query=idIndex:multiKSQ[1]@2147483647
JanusGraphVertexStep(BOTH,vertex)                                     20          20           6.600    60.34
    \_condition=(EDGE AND visibility:normal)
    \_isFitted=true
    \_vertices=1
    \_query=org.janusgraph.diskstorage.keycolumnvalue.SliceQuery@801a60ee
    \_orders=[]
    \_isOrdered=true
    \_multi=true
  optimization                                                                                 0.026
  backend-query                                                       20                       5.849
    \_query=org.janusgraph.diskstorage.keycolumnvalue.SliceQuery@801a60ee
                                            >TOTAL                     -           -          10.938        -


_**Though limit is set to 1, even then query=idIndex:multiKSQ[1]@2147483647 is using Integer.MAX_INT**_

**Reply by li-boxuan** [reply]
Can you show what the result of "mgmt.printIndexes()" is and what your JanusGraph version is?

**Reply by sauvermazeotap** [reply]
JG version is 0.5.2

PFB 

`gremlin> mgmt = graph.openManagement()
==>org.janusgraph.graphdb.database.management.ManagementSystem@a91119b
gremlin> mgmt.printIndexes()
==>------------------------------------------------------------------------------------------------
Vertex Index Name              | Type        | Unique    | Backing        | Key:           Status |
---------------------------------------------------------------------------------------------------
idIndex                        | Composite   | true      | internalindex  | idValue:      ENABLED |
                               |             |           |                | idType:       ENABLED |
---------------------------------------------------------------------------------------------------
Edge Index (VCI) Name          | Type        | Unique    | Backing        | Key:           Status |
---------------------------------------------------------------------------------------------------
---------------------------------------------------------------------------------------------------
Relation Index                 | Type        | Direction | Sort Key       | Order    |     Status |
---------------------------------------------------------------------------------------------------`

_**The output is misformatted due to github markdown**_

**Reply by sauvermazeotap** [reply]
hey @li-boxuan, my bad, PFB observation with 0.5.2 

gremlin> g.V().has('idValue','011c40f6-18d8-4111-77be-7ff5238d1eed').has('idType','id_mid_4').limit(1).bothE().otherV().profile()
==>Traversal Metrics
Step                                                               Count  Traversers       Time (ms)    % Dur
=============================================================================================================
JanusGraphStep([],[idValue.eq(011c40f6-18d8-411...                     1           1           2.747    84.89
    \_condition=(idValue = 011c40f6-18d8-4111-77be-7ff5238d1eed AND idType = id_mid_4)
    \_orders=[]
    **\_limit=1**
    \_isFitted=true
    \_isOrdered=true
    **\_query=multiKSQ[1]@4**
    \_index=idIndex
  optimization                                                                                 0.025
  optimization                                                                                 0.360
  backend-query                                                                                0.000
    \_query=idIndex:multiKSQ[1]@4
    \_limit=4
JanusGraphVertexStep(BOTH,vertex)                                     20          20           0.489    15.11
    \_condition=(EDGE AND visibility:normal)
    \_orders=[]
    \_isFitted=true
    \_isOrdered=true
    \_query=org.janusgraph.diskstorage.keycolumnvalue.SliceQuery@8009c75e
    \_multi=true
    \_vertices=1
  optimization                                                                                 0.028
                                            >TOTAL                     -           -           3.237        -

One question here : with limit 1, why limit=4 is applied in multiKSQ?

**Reply by li-boxuan** [reply]
> with limit 1, why limit=4 is applied in multiKSQ

That is a bug which is fixed in 0.6.0. The scheduled release deadline for 0.6.0 is May 1st.

**Reply by sauvermazeotap** [reply]
thanks @li-boxuan for the quick response
